File: python/agent.py
import os
import json
import asyncio
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel, Field
from dotenv import load_dotenv
from typing import List, Dict, Any, Optional
import time
import uuid

from llm import get_llm
from tools import get_filtered_tools, get_tool_definitions

# Load environment variables from a .env file
load_dotenv()

# --- Configuration ---

PROVIDER_NAME = os.getenv("PROVIDER")
MODEL_NAME = os.getenv("MODEL")

ALLOWED_TOOLS_CSV = os.getenv("ALLOWED_TOOLS")
ALLOWED_TOOLS_LIST = ALLOWED_TOOLS_CSV.split(',') if ALLOWED_TOOLS_CSV else None
MAX_TOOL_CALLS = int(os.getenv("MAX_TOOL_CALLS", 5))


def load_system_prompt(prompt_env_var: str) -> str:
    default_prompt = "You are a helpful AI assistant."
    if not prompt_env_var:
        return default_prompt
    if os.path.exists(prompt_env_var):
        try:
            with open(prompt_env_var, 'r', encoding='utf-8') as f:
                return f.read().strip()
        except Exception as e:
            print(f"Warning: Could not read system prompt file '{prompt_env_var}'. Error: {e}. Using default.")
            return default_prompt
    return prompt_env_var

SYSTEM_PROMPT = load_system_prompt(os.getenv("SYSTEM_PROMPT"))

if not PROVIDER_NAME:
    raise ValueError("PROVIDER must be set in the environment or .env file.")

# --- FastAPI App and Models ---

app = FastAPI(
    title="Multi-Provider LLM Agent API",
    description="A conversational AI agent with tool-use capabilities.",
)

class ChatMessage(BaseModel):
    role: str
    content: str

class ChatRequest(BaseModel):
    message: str
    history: List[ChatMessage] = Field(default_factory=list)

class ChatResponse(BaseModel):
    response: str
    updated_history: List[ChatMessage]

# OpenAI-compatible models
class OpenAIMessage(BaseModel):
    role: str
    content: str
    tool_calls: Optional[List[Dict[str, Any]]] = None

class OpenAIChatRequest(BaseModel):
    model: Optional[str] = None
    messages: List[OpenAIMessage]
    temperature: Optional[float] = 0.7
    max_tokens: Optional[int] = 1024
    tools: Optional[List[Dict[str, Any]]] = None
    stream: Optional[bool] = False

class OpenAIChoice(BaseModel):
    index: int
    message: OpenAIMessage
    finish_reason: str

class OpenAIUsage(BaseModel):
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int

class OpenAIChatResponse(BaseModel):
    id: str
    object: str = "chat.completion"
    created: int
    model: str
    choices: List[OpenAIChoice]
    usage: OpenAIUsage

# --- LLM Initialization and Storage ---

llm_provider = get_llm(PROVIDER_NAME, model=MODEL_NAME)
print(f"Successfully initialized LLM provider: {PROVIDER_NAME} with model: {llm_provider.model}")

# Filter tools based on environment configuration
AVAILABLE_TOOLS = get_filtered_tools(ALLOWED_TOOLS_LIST)
print(f"Agent initialized with tools: {list(AVAILABLE_TOOLS.keys())}")

conversations: Dict[str, List[Dict[str, Any]]] = {} # In-memory storage

# --- API Endpoints ---

@app.post("/chat", response_model=ChatResponse)
async def chat(request: ChatRequest):
    
    
    messages = [
        {"role": "system", "content": SYSTEM_PROMPT},
        *[msg.dict() for msg in request.history],
        {"role": "user", "content": request.message},
    ]

    try:
        for _ in range(MAX_TOOL_CALLS):
            response = await llm_provider.chat_completion(
                messages=messages,
                tools=get_tool_definitions(AVAILABLE_TOOLS),
                temperature=0.7,
                max_tokens=1024,
            )
            print(response)
            response_message = response['choices'][0]['message']
            if response_message['content']:
                messages.append(response_message) # Add LLM response to history

            if response_message.get("tool_calls"):
                print(f"[Agent] LLM requested tool call(s): {response_message['tool_calls']}")
                tool_calls = response_message["tool_calls"]
                tool_tasks = []
                for tool_call in tool_calls:
                    tool_name = tool_call['function']['name']
                    if tool_name not in AVAILABLE_TOOLS:
                        # This is a safeguard in case the LLM hallucinates a tool name
                        raise HTTPException(status_code=400, detail=f"Attempted to use unavailable tool: {tool_name}")
                    tool_to_call = AVAILABLE_TOOLS[tool_name]['execute']
                    tool_args = json.loads(tool_call['function']['arguments'])
                    tool_tasks.append(tool_to_call(**tool_args))
                
                tool_outputs = await asyncio.gather(*tool_tasks)

                for i, tool_call in enumerate(tool_calls):
                    messages.append({
                        "tool_call_id": tool_call['id'],
                        "role": "tool",
                        "name": tool_call['function']['name'],
                        "content": tool_outputs[i],
                    })
            else:
                final_response = response_message['content']
                updated_history = [msg for msg in messages if msg['role'] != 'system']
                conversations["default"] = updated_history
                return ChatResponse(response=final_response, updated_history=updated_history)

        raise HTTPException(status_code=500, detail="Exceeded maximum number of tool calls.")

    except Exception as e:
        print(f"LLM API error: {e}")
        raise HTTPException(status_code=500, detail=f"LLM API error: {str(e)}")

@app.post("/v1/chat/completions", response_model=OpenAIChatResponse)
async def openai_chat_completions(request: OpenAIChatRequest):
    """
    OpenAI-compatible chat completions endpoint.
    Allows this agent to be used as a drop-in replacement for OpenAI's API.
    """
    if request.stream:
        raise HTTPException(status_code=400, detail="Streaming is not currently supported")

    # Convert OpenAI messages to internal format
    messages = []
    for msg in request.messages:
        message_dict = {"role": msg.role, "content": msg.content}
        if msg.tool_calls:
            message_dict["tool_calls"] = msg.tool_calls
        messages.append(message_dict)

    # Use tools from request or default to available tools
    tools = request.tools if request.tools is not None else get_tool_definitions(AVAILABLE_TOOLS)

    try:
        for _ in range(MAX_TOOL_CALLS):
            response = await llm_provider.chat_completion(
                messages=messages,
                tools=tools,
                temperature=request.temperature,
                max_tokens=request.max_tokens,
            )

            response_message = response['choices'][0]['message']
            if response_message['content']:
                messages.append(response_message)

            if response_message.get("tool_calls"):
                print(f"[Agent] LLM requested tool call(s): {response_message['tool_calls']}")
                tool_calls = response_message["tool_calls"]
                tool_tasks = []
                for tool_call in tool_calls:
                    tool_name = tool_call['function']['name']
                    if tool_name not in AVAILABLE_TOOLS:
                        raise HTTPException(status_code=400, detail=f"Attempted to use unavailable tool: {tool_name}")
                    tool_to_call = AVAILABLE_TOOLS[tool_name]['execute']
                    tool_args = json.loads(tool_call['function']['arguments'])
                    tool_tasks.append(tool_to_call(**tool_args))

                tool_outputs = await asyncio.gather(*tool_tasks)

                for i, tool_call in enumerate(tool_calls):
                    messages.append({
                        "tool_call_id": tool_call['id'],
                        "role": "tool",
                        "name": tool_call['function']['name'],
                        "content": tool_outputs[i],
                    })
            else:
                # Final response - format as OpenAI response
                return OpenAIChatResponse(
                    id=f"chatcmpl-{uuid.uuid4().hex[:24]}",
                    created=int(time.time()),
                    model=request.model or llm_provider.model,
                    choices=[
                        OpenAIChoice(
                            index=0,
                            message=OpenAIMessage(
                                role="assistant",
                                content=response_message['content'],
                                tool_calls=response_message.get('tool_calls')
                            ),
                            finish_reason="stop"
                        )
                    ],
                    usage=OpenAIUsage(
                        prompt_tokens=0,  # Would need to implement token counting
                        completion_tokens=0,
                        total_tokens=0
                    )
                )

        raise HTTPException(status_code=500, detail="Exceeded maximum number of tool calls.")

    except Exception as e:
        print(f"LLM API error: {e}")
        raise HTTPException(status_code=500, detail=f"LLM API error: {str(e)}")

@app.get("/v1/models")
async def list_models():
    """
    OpenAI-compatible models endpoint.
    Returns the currently configured model.
    """
    return {
        "object": "list",
        "data": [
            {
                "id": llm_provider.model,
                "object": "model",
                "created": int(time.time()),
                "owned_by": PROVIDER_NAME,
                "permission": [],
                "root": llm_provider.model,
                "parent": None,
            }
        ]
    }

@app.get("/health")
async def health():
    return {
        "status": "healthy",
        "configured_provider": PROVIDER_NAME,
        "configured_model": llm_provider.model,
    }

if __name__ == "__main__":
    import uvicorn
    # This allows you to run the app directly with `python agent.py`
    uvicorn.run(app, host="0.0.0.0", port=8000)

File: js/agent.js
import express from 'express';
import dotenv from 'dotenv';
import fs from 'fs';
import path from 'path';
import { randomUUID } from 'crypto';
import { getLlm } from './llm.js';
import { getFilteredTools, getToolDefinitions } from './tools.js';

// --- Configuration ---
dotenv.config();

const app = express();
const port = process.env.PORT || 8000;

const PROVIDER_NAME = process.env.PROVIDER;
const MODEL_NAME = process.env.MODEL;

const ALLOWED_TOOLS_CSV = process.env.ALLOWED_TOOLS;
const ALLOWED_TOOLS_LIST = ALLOWED_TOOLS_CSV ? ALLOWED_TOOLS_CSV.split(',') : null;

function loadSystemPrompt(promptEnvVar) {
    const defaultPrompt = "You are a helpful AI assistant.";
    if (!promptEnvVar) {
        return defaultPrompt;
    }
    if (fs.existsSync(promptEnvVar)) {
        try {
            return fs.readFileSync(promptEnvVar, 'utf-8').trim();
        } catch (e) {
            console.warn(`Could not read system prompt file '${promptEnvVar}'. Error: ${e}. Using default.`);
            return defaultPrompt;
        }
    }
    return promptEnvVar;
}

const SYSTEM_PROMPT = loadSystemPrompt(process.env.SYSTEM_PROMPT);

if (!PROVIDER_NAME) {
    throw new Error("PROVIDER must be set in the environment or .env file.");
}

// --- LLM Provider Initialization ---
let llmProvider;
try {
    llmProvider = getLlm(PROVIDER_NAME, { model: MODEL_NAME });
    console.log(`Successfully initialized LLM provider: ${PROVIDER_NAME} with model: ${llmProvider.model}`);
} catch (e) {
    console.error("Failed to initialize LLM provider on startup.", e);
    process.exit(1);
}

// Filter tools based on environment configuration
const AVAILABLE_TOOLS = getFilteredTools(ALLOWED_TOOLS_LIST);
console.log(`Agent initialized with tools: ${Object.keys(AVAILABLE_TOOLS)}`);

// --- Middleware and In-memory Storage ---
app.use(express.json());
const conversations = {}; // In-memory conversation storage

// --- API Endpoints ---

app.post("/chat", async (req, res) => {
    const { message: userMessage, history: providedHistory = [] } = req.body;
    const MAX_TOOL_CALLS = 5; // Safety to prevent infinite loops

    if (!userMessage) {
        return res.status(400).json({ error: "The 'message' field is required." });
    }

    // Start building the message history for the LLM
    const messages = [
        { role: "system", content: SYSTEM_PROMPT },
        ...providedHistory,
        { role: "user", content: userMessage }
    ];

    try {
        let toolCallCount = 0;
        while (toolCallCount < MAX_TOOL_CALLS) {
            const response = await llmProvider.chatCompletion({
                messages,
                tools: getToolDefinitions(AVAILABLE_TOOLS), // Provide the list of allowed tools to the LLM
                temperature: 0.7,
                max_tokens: 1024,
            });

            const responseMessage = response.choices[0].message;

            // Check if the LLM wants to call a tool
            if (responseMessage.tool_calls && responseMessage.tool_calls.length > 0) {
                console.log("[Agent] LLM requested tool call(s):", responseMessage.tool_calls);
                messages.push(responseMessage); // Add LLM's tool request to history

                // Execute all requested tools in parallel
                const toolPromises = responseMessage.tool_calls.map(async (toolCall) => {
                    const toolName = toolCall.function.name;
                    const tool = AVAILABLE_TOOLS[toolName];

                    if (!tool) {
                        // This is a safeguard in case the LLM hallucinates a tool name
                        console.warn(`LLM attempted to use unavailable tool: ${toolName}`);
                    }

                    if (!tool) {
                        return {
                            tool_call_id: toolCall.id,
                            role: "tool",
                            name: toolName,
                            content: `Error: Tool "${toolName}" not found.`,
                        };
                    }

                    const args = JSON.parse(toolCall.function.arguments);
                    const toolResult = await tool.execute(...Object.values(args));

                    return {
                        tool_call_id: toolCall.id,
                        role: "tool",
                        name: toolName,
                        content: toolResult,
                    };
                });

                const toolResults = await Promise.all(toolPromises);
                messages.push(...toolResults); // Add tool results to history
                toolCallCount++; // Continue the loop to let the LLM synthesize the results

            } else {
                // No tool call, so this is the final answer
                const responseText = responseMessage.content;
                
                // Update and save history
                const updatedHistory = [...providedHistory, { role: "user", content: userMessage }, { role: "assistant", content: responseText }];
                conversations["default"] = updatedHistory; // Using a simple default conversation ID

                return res.json({
                    response: responseText,
                    updated_history: updatedHistory,
                });
            }
        }
        // If the loop exits, we've hit the tool call limit
        res.status(500).json({ error: "Exceeded maximum number of tool calls." });

    } catch (e) {
        console.error("LLM API error:", e);
        res.status(500).json({ error: `LLM API error: ${e.message}` });
    }
});

// OpenAI-compatible chat completions endpoint
app.post("/v1/chat/completions", async (req, res) => {
    const {
        model,
        messages,
        temperature = 0.7,
        max_tokens = 1024,
        tools = null,
        stream = false
    } = req.body;
    const MAX_TOOL_CALLS = 5;

    if (stream) {
        return res.status(400).json({ error: "Streaming is not currently supported" });
    }

    if (!messages || !Array.isArray(messages)) {
        return res.status(400).json({ error: "The 'messages' field is required and must be an array." });
    }

    // Use tools from request or default to available tools
    const toolsToUse = tools !== null ? tools : getToolDefinitions(AVAILABLE_TOOLS);

    try {
        let toolCallCount = 0;
        const workingMessages = [...messages];

        while (toolCallCount < MAX_TOOL_CALLS) {
            const response = await llmProvider.chatCompletion({
                messages: workingMessages,
                tools: toolsToUse,
                temperature,
                max_tokens,
            });

            const responseMessage = response.choices[0].message;

            // Check if the LLM wants to call a tool
            if (responseMessage.tool_calls && responseMessage.tool_calls.length > 0) {
                console.log("[Agent] LLM requested tool call(s):", responseMessage.tool_calls);
                workingMessages.push(responseMessage);

                const toolPromises = responseMessage.tool_calls.map(async (toolCall) => {
                    const toolName = toolCall.function.name;
                    const tool = AVAILABLE_TOOLS[toolName];

                    if (!tool) {
                        console.warn(`LLM attempted to use unavailable tool: ${toolName}`);
                        return {
                            tool_call_id: toolCall.id,
                            role: "tool",
                            name: toolName,
                            content: `Error: Tool "${toolName}" not found.`,
                        };
                    }

                    const args = JSON.parse(toolCall.function.arguments);
                    const toolResult = await tool.execute(...Object.values(args));

                    return {
                        tool_call_id: toolCall.id,
                        role: "tool",
                        name: toolName,
                        content: toolResult,
                    };
                });

                const toolResults = await Promise.all(toolPromises);
                workingMessages.push(...toolResults);
                toolCallCount++;

            } else {
                // Final response - format as OpenAI response
                return res.json({
                    id: `chatcmpl-${randomUUID().replace(/-/g, '').slice(0, 24)}`,
                    object: "chat.completion",
                    created: Math.floor(Date.now() / 1000),
                    model: model || llmProvider.model,
                    choices: [
                        {
                            index: 0,
                            message: {
                                role: "assistant",
                                content: responseMessage.content,
                                tool_calls: responseMessage.tool_calls || null
                            },
                            finish_reason: "stop"
                        }
                    ],
                    usage: {
                        prompt_tokens: 0,  // Would need to implement token counting
                        completion_tokens: 0,
                        total_tokens: 0
                    }
                });
            }
        }

        res.status(500).json({ error: "Exceeded maximum number of tool calls." });

    } catch (e) {
        console.error("LLM API error:", e);
        res.status(500).json({ error: `LLM API error: ${e.message}` });
    }
});

app.get("/v1/models", (req, res) => {
    res.json({
        object: "list",
        data: [
            {
                id: llmProvider.model,
                object: "model",
                created: Math.floor(Date.now() / 1000),
                owned_by: PROVIDER_NAME,
                permission: [],
                root: llmProvider.model,
                parent: null,
            }
        ]
    });
});

app.get("/health", (req, res) => {
    res.json({
        status: "healthy",
        configured_provider: PROVIDER_NAME,
        configured_model: llmProvider.model,
    });
});

// --- Server Start ---
app.listen(port, () => {
    console.log(`Server is running on http://localhost:${port}`);
});